core/trino-main/src/main/java/io/trino/execution/scheduler/FaultTolerantStageScheduler.java

```diff
@@ -0,0 +1,529 @@
+    @GuardedBy("this")
+    private final Multimap<Integer, RemoteTask> partitionToRemoteTaskMap = ArrayListMultimap.create();
+    @GuardedBy("this")
+    private final Map<TaskId, RemoteTask> runningTasks = new HashMap<>();
+    @GuardedBy("this")
+    private final Map<TaskId, InternalNode> runningNodes = new HashMap<>();
+    @GuardedBy("this")
+    private final Queue<Integer> queuedPartitions = new ArrayDeque<>();
+    @GuardedBy("this")
+    private final Set<Integer> finishedPartitions = new HashSet<>();
+    @GuardedBy("this")
+    private int remainingRetryAttempts;
+
+    @GuardedBy("this")
+    private Throwable failure;
+    @GuardedBy("this")
+    private boolean closed;
+
+    public FaultTolerantStageScheduler(
+            Session session,
+            SqlStageExecution stageExecution,
+            FailureDetector failureDetector,
+            TaskSourceFactory taskSourceFactory,
+            NodeAllocator nodeAllocator,
+            TaskLifecycleListener taskLifecycleListener,
+            Optional<Exchange> sinkExchange,
+            Optional<int[]> sinkBucketToPartitionMap,
+            Map<PlanFragmentId, Exchange> sourceExchanges,
+            Optional<int[]> sourceBucketToPartitionMap,
+            Optional<BucketNodeMap> sourceBucketNodeMap,
+            int retryAttempts)
+    {
+        checkArgument(!stageExecution.getFragment().getStageExecutionDescriptor().isStageGroupedExecution(), "grouped execution is expected to be disabled");
+
+        this.session = requireNonNull(session, "session is null");
+        this.stageExecution = requireNonNull(stageExecution, "stageExecution is null");
+        this.failureDetector = requireNonNull(failureDetector, "failureDetector is null");
+        this.taskSourceFactory = requireNonNull(taskSourceFactory, "taskSourceFactory is null");
+        this.nodeAllocator = requireNonNull(nodeAllocator, "nodeAllocator is null");
+        this.taskLifecycleListener = requireNonNull(taskLifecycleListener, "taskLifecycleListener is null");
+        this.sinkExchange = requireNonNull(sinkExchange, "sinkExchange is null");
+        this.sinkBucketToPartitionMap = requireNonNull(sinkBucketToPartitionMap, "sinkBucketToPartitionMap is null");
+        this.sourceExchanges = ImmutableMap.copyOf(requireNonNull(sourceExchanges, "sourceExchanges is null"));
+        this.sourceBucketToPartitionMap = requireNonNull(sourceBucketToPartitionMap, "sourceBucketToPartitionMap is null");
+        this.sourceBucketNodeMap = requireNonNull(sourceBucketNodeMap, "sourceBucketNodeMap is null");
+        this.remainingRetryAttempts = retryAttempts;
+    }
+
+    public StageId getStageId()
+    {
+        return stageExecution.getStageId();
+    }
+
+    public synchronized ListenableFuture<Void> isBlocked()
+    {
+        return nonCancellationPropagating(blocked);
+    }
+
+    public synchronized void schedule()
```

Comment:
given fact that this is `synchronized` cannot we get to the state when one of the threads will be executing `while` loop inside. And we will have multiple threads waiting on executor trying to execute another instance of `schedule()`?

Reply:
Maybe we can ensure that this class is always called out by single thread?